### DISCUSSION TITLE
Android: ClassNotFoundException for controls in a library

### DISCUSSION THREAD

**Discussion by Arlodotexe** [unoplatform/uno#8345] [Q&A]
### Current behavior

When using a backing library for controls (Page, UserControl, etc), trying to instantiate the control on Android is causing a `Java.Lang.ClassNotFoundException`

> [monodroid] JNIEnv.FindClass(Type) caught unexpected exception: Java.Lang.ClassNotFoundException: crc64dc3abbcce329d998.TestUserControl ---> Java.Lang.ClassNotFoundException: Didn't find class "crc64dc3abbcce329d998.TestUserControl" on path: DexPathList[[zip file "/data/app/~~4OtF9KiBiPY8WK683yGzVw==/ClassNotFoundExceptionDroid.ClassNotFoundExceptionDroid-tVHNMKEExLiFkxxvWa21Vw==/base.apk"],nativeLibraryDirectories=[/data/app/~~4OtF9KiBiPY8WK683yGzVw==/ClassNotFoundExceptionDroid.ClassNotFoundExceptionDroid-tVHNMKEExLiFkxxvWa21Vw==/lib/x86_64, /data/app/~~4OtF9KiBiPY8WK683yGzVw==/ClassNotFoundExceptionDroid.ClassNotFoundExceptionDroid-tVHNMKEExLiFkxxvWa21Vw==/base.apk!/lib/x86_64, /system/lib64, /system_ext/lib64]]

![image](https://user-images.githubusercontent.com/9384894/157994723-3aaccac0-e452-45cd-a1f6-0203bcd6ced2.png)

### Expected behavior

No exception should throw.

### How to reproduce it (as minimally and precisely as possible)

1. Create a `dotnet new unoapp`
2. Add an "Uno Class Library" to the solution
3. Add a reference to this library to your Droid head.
4. Create a user control in the library
5. Use the UserControl in the shared project (either XAML or manually instantiated from C#)
6. Run on Android and observe the error.

Or, run this premade solution:
[ClassNotFoundExceptionDroid.zip](https://github.com/unoplatform/uno/files/8236628/ClassNotFoundExceptionDroid.zip)

### Works on UWP/WinUI

Yes

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia

### NuGet package version(s)

Uno.UI latest stable 4.1.8

### Affected platforms

Android

### IDE

Visual Studio 2022

### IDE version

17.0.5

**Comment by jeromelaban** [comment]
Thanks for the report. In the sample, the class library provides `net6.0-*` and `netstandard2.0` target frameworks, which is then referenced from a xamarin app.

As Xamarin (non-net6 ones) apps don't support net6, the build system ends up using `netstandard2.0` which is only providing wasm+skia support. Android tries to find is support content for controls native counterparts, but those don't exist as they've not been generated at compile time.

To fix this, add `monoandroid12.0` and `xamarinios10.0` in the target frameworks, like this:
https://github.com/unoplatform/uno/blob/d2e1e454f52e82cd31da2719b5d1aad2a4c8c9e8/src/SolutionTemplate/UnoLibraryTemplate/CrossTargetedLibrary.csproj#L8
